Drop StripRedundantLastElementsOfPath() function

hebasto opened the pull request:
[Switching](https://github.com/bitcoin/bitcoin/pull/20744) to `std::filesystems` makes possible to leverage [`std::filesystem::path::lexically_normal`](https://en.cppreference.com/w/cpp/filesystem/path/lexically_normal) and get rid of ugly `StripRedundantLastElementsOfPath()` crutch.

To make its usage simple and error-proof, a new `ArgsManager::GetPathArg()` member function introduced which guarantees to return a normalized with no trailing slashes paths provided via `-datadir`, `-blocksdir` or `-walletdir` command-line arguments or configure options.

hebasto committed:
util: Add ArgsManager::GetPathArg() function

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

hebasto committed:
Use ArgsManager::GetPathArg() for "-datadir" option

hebasto committed:
Use ArgsManager::GetPathArg() for "-blocksdir" option

hebasto committed:
Use ArgsManager::GetPathArg() for "-walletdir" option

hebasto committed:
util: Drop no longer needed StripRedundantLastElementsOfPath() function

ryanofsky commented:
> Also it became possible to re-enable Windows tests which were disabled in #20744. So this PR, apparently, is an alternative to the #24251.

This all sounds good. IMO, it would be good to merge #24251 before this PR. #24251 has two very simple fixes for the regressions caused by #20744 yesterday. This PR should be trivial to rebase if #24251 is merged first, and should become a little simpler, too.

ryanofsky commented on src/util/system.h:
```diff
@@ -264,6 +264,16 @@ class ArgsManager
+     * @param pathlike_arg Pathlike argument to get a path from (e.g., "-datadir", "-blocksdir" or "-walletdir")
+     * @return Normalized path which is get from a specified pathlike argument
+     */
+    fs::path GetNormalPath(std::string pathlike_arg) const;
```
In commit "util: Add ArgsManager::GetNormalPath() function" (034f12a115e21b4205b532a3f946b129e83505d9)

Minor: Would vote for calling this `GetPathArg` to be consistent with `GetIntArg` `GetBoolArg`, instead of `GetNormalPath`

hebasto replied:
Thanks! [Done](https://github.com/bitcoin/bitcoin/pull/24265#issuecomment-1030404670).

ryanofsky commented on src/util/system.cpp:
```diff
@@ -399,6 +399,15 @@ std::optional<unsigned int> ArgsManager::GetArgFlags(const std::string& name) co
+    if (path_str.empty()) return {};
+    // Adding a slash followed by parent_path() call
+    // guaranties no trailing slashes in the resulted path.
+    return fs::PathFromString(path_str + "/").lexically_normal().parent_path();
```
In commit "util: Add ArgsManager::GetNormalPath() function" (034f12a115e21b4205b532a3f946b129e83505d9)

I think previous approach which avoids direct string manipulation and avoids assuming "/" is a path separator is a little cleaner. Would suggest something like:

```c++
fs::path result = fs::PathFromString(path_str).lexically_normal(); // Collapse slashes and dots
return result.has_filename() ? result : result.parent_path(); // Remove trailing slash, if present
```

Also it would be great to write a simple unit test for this new function.

hebasto replied:
Thanks! [Done](https://github.com/bitcoin/bitcoin/pull/24265#issuecomment-1030623898).

hebasto commented:
@ryanofsky 

Thanks for your suggestions! Most of them are implemented.

> Also it would be great to write a simple unit test for this new function.

Will add in the morning :)

fanquake commented:
> This all sounds good. IMO, it would be good to merge https://github.com/bitcoin/bitcoin/pull/24251 before this PR.

I agree, we'll merge #24251 first. Maybe rebase on top?

hebasto commented:
Updated e69ab379283af554d60c484d5715112b887fa0c3 -> 05c8d8cd075c37f43ad195bceac3eaa1f2907e07 ([pr24265.02](https://github.com/hebasto/bitcoin/commits/pr24265.02) -> [pr24265.03](https://github.com/hebasto/bitcoin/commits/pr24265.03)):

- rebased on top of the #24251
- unit test added

prusnak commented:
There are couple of more places where we can replace `fs::PathFromString(args.GetArg(...))` with `args.GetPathArg`:

```
src/init.cpp:138:    return AbsPathForConfigVal(fs::PathFromString(args.GetArg("-pid", BITCOIN_PID_FILENAME)));
src/init.cpp:1249:            fs::path asmap_path = fs::PathFromString(args.GetArg("-asmap", ""));
src/util/system.cpp:391:    auto result = fs::PathFromString(GetArg(pathlike_arg, "")).lexically_normal();
src/init/common.cpp:84:    LogInstance().m_file_path = AbsPathForConfigVal(fs::PathFromString(args.GetArg("-debuglogfile", DEFAULT_DEBUGLOGFILE)));
src/bench/bench_bitcoin.cpp:112:    args.output_csv = fs::PathFromString(argsman.GetArg("-output_csv", ""));
src/bench/bench_bitcoin.cpp:113:    args.output_json = fs::PathFromString(argsman.GetArg("-output_json", ""));
```

ryanofsky commented on src/test/getarg_tests.cpp:
```diff
@@ -159,6 +159,98 @@ BOOST_AUTO_TEST_CASE(intarg)
     BOOST_CHECK_EQUAL(m_local_args.GetIntArg("-bar", 11), 0);
 }
 
+BOOST_AUTO_TEST_CASE(patharg)
```
In commit "util: Add ArgsManager::GetPathArg() function" (d32f6c0dc89f1a609d59a222b09eabb1eb0da542)

Nice tests!

ryanofsky commented:
> There are couple of more places where we can do similar optimization, but for arguments which expect filename, not path (= directory).
>
> It would be possible to introduce a `GetFileArg` function using the following refactor and use that to simplify the lines above:

I'd push back on these suggestions a little and save these things for followups. They would be natural extensions of this PR, but would risk unintentionally changing behavior and don't need to be part of it. Also a `std::string` default argument would pass more paths as strings, when I think paths should be represented by `fs::path` not `std::string` wherever possible.

ryanofsky reviewed (approved):
Code review ACK 05c8d8cd075c37f43ad195bceac3eaa1f2907e07. This looks great, and I really like the way you broke it up into isolated commits. Make this very easy to review.

re: https://github.com/bitcoin/bitcoin/pull/24265#issue-1124619834 

> Also it became possible to re-enable Windows tests which were disabled in #20744. So this PR, apparently, is an alternative to the #24251.

This is a little out of date and could be removed from the description

hebasto commented:
> > There are couple of more places where we can do similar optimization, but for arguments which expect filename, not path (= directory).
> > It would be possible to introduce a `GetFileArg` function using the following refactor and use that to simplify the lines above:
> 
> I'd push back on these suggestions a little and save these things for followups. They would be natural extensions of this PR, but would risk unintentionally changing behavior and don't need to be part of it. Also a `std::string` default argument would pass more paths as strings, when I think paths should be represented by `fs::path` not `std::string` wherever possible.

Agree.

hebasto commented:
> > Also it became possible to re-enable Windows tests which were disabled in #20744. So this PR, apparently, is an alternative to the #24251.
> 
> This is a little out of date and could be removed from the description

PR description has been updated.

prusnak commented:
> I'd push back on these suggestions a little and save these things for followups.

Makes sense. Created a draft PR https://github.com/bitcoin/bitcoin/pull/24274 which builds on top of this PR.

hebasto commented:
@prusnak 
> > I'd push back on these suggestions a little and save these things for followups.
> 
> 
> 
> Makes sense. Created a draft PR https://github.com/bitcoin/bitcoin/pull/24274 which builds on top of this PR.

Mind leaving (N)ACK review comment?

prusnak commented on src/util/system.h:
```diff
@@ -264,6 +264,16 @@ class ArgsManager
+     * @param pathlike_arg Pathlike argument to get a path from (e.g., "-datadir", "-blocksdir" or "-walletdir")
+     * @return Normalized path which is get from a specified pathlike argument
+     */
+    fs::path GetPathArg(std::string pathlike_arg) const;
```
Other `Get...Arg` methods use the camelCase notation for arguments. Let's change to `pathlikeArg` or similar.

hebasto replied:
For new code we strive to follow our [symbol naming convention](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c):
> Variable (including function arguments) and namespace names are all lowercase and may use _ to separate words (snake_case).

prusnak commented on src/util/system.cpp:
```diff
@@ -399,6 +386,13 @@ std::optional<unsigned int> ArgsManager::GetArgFlags(const std::string& name) co
     return std::nullopt;
 }
 
+fs::path ArgsManager::GetPathArg(std::string pathlike_arg) const
```
Other `Get...Arg` methods use the camelCase notation for arguments. Let's change to `pathlikeArg` or similar.

hebasto replied:
https://github.com/bitcoin/bitcoin/pull/24265#discussion_r800153360

prusnak reviewed (approved):
ACK 05c8d8c

MarcoFalke commented:
This is "refactoring"?

hebasto commented:
@MarcoFalke 
> This is "refactoring"?

There are some changes, I assume they are nice, in behavior. For example, the following works:
```
$ ./src/bitcoind -datadir=/root/../home/hebasto/.bitcoin
```

laanwj commented:
Concept ACK.

hebasto commented:
Rebased 05c8d8cd075c37f43ad195bceac3eaa1f2907e07 -> ebda2b8c819d989327c6b3e29237dfb43628e647 ([pr24265.03](https://github.com/hebasto/bitcoin/commits/pr24265.03) -> [pr24265.04](https://github.com/hebasto/bitcoin/commits/pr24265.04)) due to the conflict with #24266.

ryanofsky reviewed (approved):
Code review ACK ebda2b8c819d989327c6b3e29237dfb43628e647. Only change since last review is rebase which simplified the last commit

prusnak commented:
Since this PR has been merged to master, I rebased https://github.com/bitcoin/bitcoin/pull/24274 and marked it as ready for review.
